flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/SingleSessionManager.java

```diff
@@ -179,6 +186,12 @@ protected ResultFetcher callRemoveJar(OperationHandle operationHandle, String ja
                                         .getUserClassLoader())
                         .removeURL(jarURL);
             }
+            if (jars.contains(jarPath)) {
+                throw new UnsupportedOperationException(
+                        String.format(
+                                "Can not delete jar file [%s] while it is being used by function.",
+                                jarPath));
+            }
```

Comment:
The jar list that is returned by the `SHOW JAR` syntax also contains the connector jars. It's a little hard to check if the jar is a connector jar or a function jar. If it is a connector jar, I think we can not check whether we have instances using this jar.